### DISCUSSION TITLE
SemVer type

### DISCUSSION THREAD

**Discussion by Kludex** [pydantic/pydantic#2506] [Feature Request]
What do you think about a `SemVer` type?

Usage:
```python
class Potato(BaseModel):
    version: SemVer

potato = Potato(version="1.0.0")
```

**Comment by benipeled** [comment]
+1

Meanwhile, I found this implementation - https://python-semver.readthedocs.io/en/latest/advanced/combine-pydantic-and-semver.html

```python
from semver import VersionInfo
from pydantic import BaseModel


class SemVer(VersionInfo):
    @classmethod
    def __get_validators__(cls):
        yield cls.parse

    @classmethod
    def __modify_schema__(cls, field_schema):
        field_schema.update(examples=["1.0.2", "2.15.3-alpha", "21.3.15-beta+12345"])


class MyModel(BaseModel):
    version: SemVer


MyModel(version='1.2.3')
```

Verified with semver==2.13.0

**Comment by jbkroner** [comment]
I would like to pick this up as it would help a project I'm working on and seems like a fun first contribution.

I see two paths to implementing this: 
- self contained `SemVer` type in Pydantic, not reliant on the `semver` package
- add the `semver` package as a dependency and construct it similarly to example above / in the `semver` docs.  This seems like the highest payoff / effort ratio.

Is there any reason we cannot add that package as a dependency to [pydantic-extra-types](https://github.com/pydantic/pydantic-extra-types)?  

edit: A slightly updated example from above, this supports pydantic >= 2.0

edit 2: need to mess around a bit with the __get_validators__ method but you get the idea.
```python
from semver import Version
from pydantic import basemodel

# define a SemVer type
class SemVer(Version):
    """Represents a Semantic Versioning (SemVer)"""

    @classmethod
    def __get_validators__(cls):
        yield cls.parse

    @classmethod
    def __get_pydantic_json_schema__(cls, field_schema):
        field_schema.update(type="string")
        
# using our new type in a pydantic model:
class myModel(BaseModel):
    version: SemVer

# usage
m = myModel(version='1.2.3')
m.version 
>>> 1.2.3
m.version.major
>>> 1
```

**Reply by jbkroner** [reply]
https://github.com/pydantic/pydantic-extra-types/pull/199

**Comment by gaardhus** [comment]
How do you deal with dev versions, i.e. `0.9.1.dev66+gd637b9609`?
